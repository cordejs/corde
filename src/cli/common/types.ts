import { Command } from "commander";

export type ParamsFrom = "args" | "options" | "both";

export interface ICommandOptions {
  flags: string;
  description?: string;
  defaultValue?: string;
}

export interface ICliCommandConstructor {
  name?: string;
  program: Command;
  paramsFrom?: ParamsFrom;
}

export interface ICliCommand {
  readonly paramsFrom?: ParamsFrom;
  readonly command: Command;
  handler(...args: any[]): Promise<void> | void;
}

export interface ICliCommandConstructable {
  new (command: Command): ICliCommand;
}
