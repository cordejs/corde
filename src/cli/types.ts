export interface ICommand {
  action(...args: any[]): Promise<void> | void;
}
