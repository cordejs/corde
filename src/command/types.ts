import { ICordeBot } from "../types";

export interface ICreateMatcherParam {
  matcher?: string;
  isNot?: boolean;
  commandName?: string | boolean | number;
  channelId?: string;
  guildId?: string;
  isDebug?: boolean;
  isCascade?: boolean;
  cordeBot?: ICordeBot;
  trace?: string;
  utility?: boolean;
  mustSendCommand?: boolean;
}
