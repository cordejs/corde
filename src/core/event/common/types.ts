import { Client, ClientEvents } from "discord.js";

export type EventNames = keyof ClientEvents;

export type Spread<T extends EventNames> = ClientEvents[T][1] extends undefined
  ? ClientEvents[T] extends (infer U)[]
    ? U
    : never
  : ClientEvents[T];

export interface IDiscordEvent<TEvent extends EventNames, TFilter> {
  canListen(): boolean;
  getIntents(): string[];
  on(fn: (...args: ClientEvents[TEvent]) => void): this;
  once(filter?: TFilter): Promise<Spread<TEvent>>;
}

export interface IDiscordEventConstructable<TEvent extends EventNames, TFilter> {
  new (client: Client): IDiscordEvent<TEvent, TFilter>;
}

export type IntentType =
  | "GUILDS"
  | "GUILD_MEMBERS"
  | "GUILD_BANS"
  | "GUILD_EMOJIS_AND_STICKERS"
  | "GUILD_INTEGRATIONS"
  | "GUILD_WEBHOOKS"
  | "GUILD_INVITES"
  | "GUILD_VOICE_STATES"
  | "GUILD_PRESENCES"
  | "GUILD_MESSAGES"
  | "GUILD_MESSAGE_REACTIONS"
  | "GUILD_MESSAGE_TYPING"
  | "DIRECT_MESSAGES"
  | "DIRECT_MESSAGE_REACTIONS"
  | "DIRECT_MESSAGE_TYPING"
  | "GUILD_SCHEDULED_EVENTS";

export type EventClass = {
  [k in IntentType]: {
    bitfield: number;
    events: EventNames[];
  };
};
