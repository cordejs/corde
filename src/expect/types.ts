declare namespace corde {
  interface IMatchers<T extends any> {
    toBe(expected: T): void;
    toEqual(expected: T): void;
    toBeCloseTo(expected: T, precisionopt: number): void;
    toBeDefined(): void;
    toBeFalse(): void;
    toBeFalsy(): void;
    toBeGreaterThan(expected: number | bigint): void;
    toBeGreaterThanOrEqual(expected: number | bigint): void;
    toBeInstanceOf<U extends any>(expected: T | U): void;
    toBeLessThan(expected: number | bigint): void;
    toBeLessThanOrEqual(expected: number | bigint): void;
    toBeNaN(): void;
    toBeNegativeInfinity(): void;
    toBeNull(): void;
    toBeUndefined(): void;
    toBePositiveInfinity(): void;
    toBeTrue(): void;
    toBeTruthy(): void;
    toContain(expected: T): void;
    toHaveBeenCalled(): void;
    toHaveBeenCalledBefore(expected): void;
    toHaveBeenCalledOnceWith(): void;
    toHaveBeenCalledTimes(amount: number): void;
    toHaveBeenCalledWith(): void;
    toHaveClass(expected): void;
    toHaveSize(size: number): void;
    toMatch(expected): void;
    toThrow<U extends any>(expectedThrow: U): void;
    toThrowError<U extends Error>(expectedopt: U): void;
    toThrowMatching(predicate): void;
    any(constructor: any): void;
    anything(mixed: any): void;
    arrayContaining(mixed): void;
    objectContaining(mixed): void;
    stringMatching(pattern): void;
    toBeArrayOfBooleans(): void;
    toBeArrayOfNumbers(): void;
    toBeArrayOfObjects(): void;
    toBeArrayOfSize(number): void;
    toBeArrayOfStrings(): void;
    toBeEmptyArray(): void;
    toBeNonEmptyArray(): void;
    toBeAfter(otherDate: Date): void;
    toBeBefore(otherDate: Date): void;
    toBeDate(): void;
    toBeValidDate(): void;
    toBeFunction(): void;
    toThrowAnyError(): void;
    toThrowErrorOfType(constructorName): void;
    toBeCalculable(): void;
    toBeEvenNumber(): void;
    toBeGreaterThanOrEqualTo(otherNumber: number | bigint): void;
    toBeLessThanOrEqualTo(otherNumber: number | bigint): void;
    toBeNear(otherNumber: number | bigint, epsilon): void;
    toBeOddNumber(): void;
    toBeWholeNumber(): void;
    toBeWithinRange(floor, ceiling): void;
    toBeEmptyObject(): void;
    toBeNonEmptyObject(): void;
    toHaveArray(memberName): void;
    toHaveArrayOfBooleans(memberName): void;
    toHaveArrayOfNumbers(memberName): void;
    toHaveArrayOfObjects(memberName): void;
    toHaveArrayOfSize(memberName, size): void;
    toHaveArrayOfStrings(memberName): void;
    toHaveBoolean(memberName): void;
    toHaveCalculable(memberName): void;
    toHaveDate(memberName): void;
    toHaveDateAfter(memberName, date): void;
    toHaveDateBefore(memberName, date): void;
    toHaveEmptyArray(memberName): void;
    toHaveEmptyObject(memberName): void;
    toHaveEmptyString(memberName): void;
    toHaveEvenNumber(memberName): void;
    toHaveFalse(memberName): void;
    toHaveHtmlString(memberName): void;
    toHaveIso8601(memberName): void;
    toHaveJsonString(memberName): void;
    toHaveMember(memberName): void;
    toHaveMethod(memberName): void;
    toHaveNonEmptyArray(memberName): void;
    toHaveNonEmptyObject(memberName): void;
    toHaveNonEmptyString(memberName): void;
    toHaveNumber(memberName): void;
    toHaveNumberWithinRange(memberName, floor, ceiling): void;
    toHaveObject(memberName): void;
    toHaveOddNumber(memberName): void;
    toHaveString(memberName): void;
    toHaveStringLongerThan(memberName: keyof T, string): void;
    toHaveStringSameLengthAs(memberName: keyof T, string): void;
    toHaveStringShorterThan(memberName: keyof T, string): void;
    toHaveTrue(memberName: keyof T): void;
    toHaveUndefined(memberName: keyof T): void;
    toHaveWhitespaceString(memberName: keyof T): void;
    toHaveWholeNumber(memberName: keyof T): void;
    toBeRegExp(): void;
    toBeEmptyString(): void;
    toBeHtmlString(): void;
    toBeIso8601(): void;
    toBeJsonString(): void;
    toBeLongerThan(otherString: string): void;
    toBeNonEmptyString(): void;
    toBeSameLengthAs(otherString: string): void;
    toBeShorterThan(otherString: string): void;
    toBeString(): void;
    toBeNumber(): void;
    toBeArray(): void;
    toBeObject(): void;
    toLength(length: number): void;
    toBeBoolean(): void;
    toBeWhitespace(): void;
    toEndWith(substring: string): void;
    toStartWith(substring: string): void;
  }

  interface Expect {
    expect<T extends any>(expect: T): corde.IMatchers<T>;
  }
}
