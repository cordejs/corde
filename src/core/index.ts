export * from "./Group";
export * from "./TestFile";
export * from "./Config";
export * from "./debug";
export * from "./Events";
export * from "./injectGlobals";
export * from "./internalEvents";
export * from "./printHookError";
export * from "./Reader";
export * from "./runtime";
export * from "./summary";
export * from "./testCollector";
export * from "./TestExecutor";
export * from "./tsRegister";
export * from "./CordeBot";
