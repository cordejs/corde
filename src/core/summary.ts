/* eslint-disable no-console */
import { TEXT_EMPTY, TEXT_FAIL, TEXT_PASS } from "../const";
import { IRunnerReport } from "../types";

interface ISummaryLine {
  label: string;
  total: number;
  fail: number;
  success: number;
  totalEmpty: number;
}

class Summary {
  print(runnerReport: IRunnerReport) {
    let message = "\n";
    message += this.buildTestFilesSummary(runnerReport) + "\n";
    message += this.buildTestsSummary(runnerReport) + "\n";
    if (runnerReport.testTimer) {
      message += "Time:       " + runnerReport.testTimer;
    }
    console.log(message);
    return message;
  }

  private buildTestFilesSummary(runnerReport: IRunnerReport) {
    return this.buildSummaryLine({
      fail: runnerReport.totalTestFilesFailed,
      label: "Test Files: ",
      success: runnerReport.totalTestFilesPassed,
      totalEmpty: runnerReport.totalEmptyTestFiles,
      total: runnerReport.totalTestFiles,
    });
  }

  private buildTestsSummary(runnerReport: IRunnerReport) {
    return this.buildSummaryLine({
      fail: runnerReport.totalTestsFailed,
      label: "Tests:      ",
      success: runnerReport.totalTestsPassed,
      totalEmpty: runnerReport.totalEmptyTests,
      total: runnerReport.totalTests,
    });
  }

  private buildSummaryLine(data: ISummaryLine) {
    const { label, fail, success, total, totalEmpty } = data;

    const partialLine: string[] = [];
    if (success > 0) {
      partialLine.push(TEXT_PASS(`${success} passed`));
    }

    if (fail > 0) {
      partialLine.push(TEXT_FAIL(`${fail} failed`));
    }

    if (totalEmpty > 0) {
      partialLine.push(TEXT_EMPTY(`${totalEmpty} empty`));
    }

    if (partialLine.length) {
      return `${label}${this.buildSummaryLinePhrase(partialLine)}. ${total} total`;
    }

    if (total === 0) {
      return `${label}${TEXT_EMPTY(`${total} total`)}`;
    }
    return `${label}${total} total`;
  }

  /**
   * Builds the summary line phrase
   *
   * @example
   *
   * ["1 passed", "2 fail"] => '1 passed and 2 fail'
   * ["1 passed", "2 fail", "1 empty"] => '1 passed, 2 fail and 1 empty'
   */
  private buildSummaryLinePhrase(parts: string[]) {
    let builtLine = "";

    for (let i = 0; i < parts.length; i++) {
      if (parts.length > 2 && i < 2) {
        builtLine += parts[i] + ", ";
      } else {
        builtLine += parts[i] + " and ";
      }
    }

    // removes the last 'and '
    return builtLine.substring(0, builtLine.length - 5);
  }
}

const summary = new Summary();
export { summary };
