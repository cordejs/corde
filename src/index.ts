export { group, test, group as describe, test as it } from "./closures";
export { expect } from "./expect";
export { sendMessage, getRole, createRole } from "./api-utilities";
export { afterAll, beforeStart, afterEach, beforeEach } from "./hooks";

export { RolePermission, Permission } from "./utils/permission";
export { Colors } from "./utils/colors";

import { corde } from "./corde";
export default corde;
