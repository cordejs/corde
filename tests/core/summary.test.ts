import { summary } from "../../src/core/summary";
import { IRunnerReport } from "../../src/types";
import { removeANSIColorStyle } from "../testHelper";

it("should print a summary for 1 test file, 1 test case, and 1 test function that has pass", () => {
  const summaryData: IRunnerReport = {
    testTimer: "2s",
    totalTestFiles: 1,
    totalTestFilesFailed: 0,
    totalTestFilesPassed: 1,
    totalTests: 1,
    totalTestsFailed: 0,
    totalTestsPassed: 1,
    totalEmptyTestFiles: 1,
    totalEmptyTests: 1,
  };

  const message = summary.print(summaryData);
  expect(removeANSIColorStyle(message)).toMatchSnapshot();
});

it("should print a summary for 1 test file, 1 test case, and 1 test function that has pass, and 1 that don't", () => {
  const summaryData: IRunnerReport = {
    testTimer: "2s",
    totalTestFiles: 1,
    totalTestFilesFailed: 0,
    totalTestFilesPassed: 1,
    totalTests: 2,
    totalTestsFailed: 1,
    totalTestsPassed: 1,
    totalEmptyTestFiles: 0,
    totalEmptyTests: 0,
  };

  const message = summary.print(summaryData);
  expect(removeANSIColorStyle(message)).toMatchSnapshot();
});

it("should print a summary for 10 test file, 10 test case, and 10 test functions", () => {
  const summaryData: IRunnerReport = {
    testTimer: "2s",
    totalTestFiles: 10,
    totalTestFilesFailed: 0,
    totalTestFilesPassed: 10,
    totalTests: 10,
    totalTestsFailed: 0,
    totalTestsPassed: 10,
    totalEmptyTestFiles: 0,
    totalEmptyTests: 0,
  };

  const message = summary.print(summaryData);
  expect(removeANSIColorStyle(message)).toMatchSnapshot();
});

it("should print a summary tests with timer ''", () => {
  const summaryData: IRunnerReport = {
    testTimer: "",
    totalTestFiles: 10,
    totalTestFilesFailed: 0,
    totalTestFilesPassed: 10,
    totalTests: 10,
    totalTestsFailed: 0,
    totalTestsPassed: 10,
    totalEmptyTestFiles: 0,
    totalEmptyTests: 0,
  };

  const message = summary.print(summaryData);
  expect(removeANSIColorStyle(message)).toMatchSnapshot();
});

it("should print a summary tests with fail, success and empty", () => {
  const summaryData: IRunnerReport = {
    testTimer: "10.658s",
    totalTestFiles: 3,
    totalTestFilesFailed: 1,
    totalTestFilesPassed: 1,
    totalTests: 3,
    totalTestsFailed: 1,
    totalTestsPassed: 1,
    totalEmptyTestFiles: 1,
    totalEmptyTests: 1,
  };

  const message = summary.print(summaryData);
  expect(removeANSIColorStyle(message)).toMatchSnapshot();
});

it("should print a summary tests with only fail", () => {
  const summaryData: IRunnerReport = {
    testTimer: "10.658s",
    totalTestFiles: 3,
    totalTestFilesFailed: 3,
    totalTestFilesPassed: 0,
    totalTests: 3,
    totalTestsFailed: 3,
    totalTestsPassed: 0,
    totalEmptyTestFiles: 0,
    totalEmptyTests: 0,
  };

  const message = summary.print(summaryData);
  expect(removeANSIColorStyle(message)).toMatchSnapshot();
});

it("should print a summary tests with only empty", () => {
  const summaryData: IRunnerReport = {
    testTimer: "10.658s",
    totalTestFiles: 1,
    totalTestFilesFailed: 0,
    totalTestFilesPassed: 0,
    totalTests: 1,
    totalTestsFailed: 0,
    totalTestsPassed: 0,
    totalEmptyTestFiles: 1,
    totalEmptyTests: 1,
  };

  const message = summary.print(summaryData);
  expect(removeANSIColorStyle(message)).toMatchSnapshot();
});

it("should print summary with only one file that is empty", () => {
  const summaryData: IRunnerReport = {
    testTimer: "3ms",
    totalTests: 0,
    totalTestFiles: 1,
    totalTestsPassed: 0,
    totalTestsFailed: 0,
    totalTestFilesPassed: 0,
    totalTestFilesFailed: 0,
    totalEmptyTestFiles: 1,
    totalEmptyTests: 0,
  };
  const message = summary.print(summaryData);
  expect(removeANSIColorStyle(message)).toMatchSnapshot();
});
