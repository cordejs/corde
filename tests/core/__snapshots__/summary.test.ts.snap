// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should print a summary for 1 test file, 1 test case, and 1 test function that has pass 1`] = `
"
Test Files: 1 passed and 1 empty. 1 total
Tests:      1 passed and 1 empty. 1 total
Time:       2s"
`;

exports[`should print a summary for 1 test file, 1 test case, and 1 test function that has pass, and 1 that don't 1`] = `
"
Test Files: 1 passed. 1 total
Tests:      1 passed and 1 failed. 2 total
Time:       2s"
`;

exports[`should print a summary for 10 test file, 10 test case, and 10 test functions 1`] = `
"
Test Files: 10 passed. 10 total
Tests:      10 passed. 10 total
Time:       2s"
`;

exports[`should print a summary tests with fail, success and empty 1`] = `
"
Test Files: 1 passed, 1 failed, 1 empty. 3 total
Tests:      1 passed, 1 failed, 1 empty. 3 total
Time:       10.658s"
`;

exports[`should print a summary tests with only empty 1`] = `
"
Test Files: 1 empty. 1 total
Tests:      1 empty. 1 total
Time:       10.658s"
`;

exports[`should print a summary tests with only fail 1`] = `
"
Test Files: 3 failed. 3 total
Tests:      3 failed. 3 total
Time:       10.658s"
`;

exports[`should print a summary tests with timer '' 1`] = `
"
Test Files: 10 passed. 10 total
Tests:      10 passed. 10 total
"
`;

exports[`should print summary with only one file that is empty 1`] = `
"
Test Files: 1 empty. 1 total
Tests:      0 total
Time:       3ms"
`;
